### ISSUE TITLE
Terrible performances of BigInteger / BigInt 

### ISSUE THREAD

**Issue by ndesmoul** [scala-js/scala-js#1785]
I did some timing comparison between the new BigInteger implementation and the jsbn javascript library. And it appears that the BigInteger implementation performances are terrible.

As example, to do a modular exponentiation with 1024-bits numbers:
jsbn (called using a Scala wrapper): 21 ms
scala BigInteger: 857 ms

With a modular multiplication (a multiply b mod n) done 100 times:
jsbn: 6 ms
scala BigInteger: 178 ms

At least the results are identical (and correct!). But there's more than a 40 factor which is pretty big.
These timings were done under Chromium, but Firefox offers similar results.

I don't know where the problem comes from. The original Java code seems to have decent performances on a standard JVM (compared to the GPL BigInteger in the standard lib). I didn't test the scala converted code under a JVM though. 
Maybe the resulting javascript code is too complicated to be correctly optimised by the javascript engines?

But as it is, the current BigInteger is not a good choice if you need to do some heavy computation involving big integers.

For a javascript implementation, jsbn offer decent performances (still at least 10 times slower than computation on a JVM...). But it is not even the fastest, as asmcrypto.js claim to be even faster (using some asm.js magic).

**Comment by ndesmoul** [comment]
Note that the timings where done using ~ fastOptsJS.
Using fullOptJS ameliorates a little the situation. The factor gap is then "only" 30. Still no good.

**Comment by sjrd** [comment]
Well ... PR welcome, I suppose.

**Comment by InTheNow** [comment]
I have a small example test, that can also run in the repl:

```scala
      import java.math.BigInteger
      import java.util.Arrays

      import java.util.Random
      val bi1 = new BigInteger(1024, new Random())
      val bi2 = new BigInteger(1024, new Random())
      val bi3 = new BigInteger(1024, new Random())
      val t1 = scala.compat.Platform.currentTime

      val b = bi1.modPow(bi2,bi3)
      val t2 = scala.compat.Platform.currentTime - t1
      println (s"$b ModPow TIME:: ${t2}ms")

      val t3 = scala.compat.Platform.currentTime
      for(i <- 0 until 100){
        val c = bi1.multiply(bi2).mod(bi3)
      }
      val t4 = scala.compat.Platform.currentTime - t3
      println (s"$b Mult mod TIME:: ${t4}ms")
```

Typical times SJS/JVM are 1000,146 / 133/129

So, on the second test the SJS is marginally slower but the first is _much_ slower. Also, the JVM test here  is way slower than the jsbn ones.

Looking at the code, jsbn uses a completely different algorithm for `ModPow`,  SJS uses the sliding window method. No doubt the JVM algorithm also differs from the SJS one. 

So I guess the solution here is to modify the actual algorithm used

**Comment by ndesmoul** [comment]
I'm not sure to fully understand your bench. When you write SJS/JVM, do 
you mean that you compare the execution time of the Scala/BigInteger 
with the standard BigInteger, both executed on a JVM?

The BigInteger converted into Scala use a sliding window technique (when 
manipulating an odd modulus), and in internal use a Montgomery 
representation.
It seems to be also the case for jsbn. It is not documented, but the 
modpow algorithm look like a window technique. And some Montgomery 
representation seems to be in use too.

And even if the Scala version used a basic square and multiply, the 
ratio should not be greater than 2 or 3 compared to a sliding window or 
a wNaf.

The problem I think is not at the modpow level. At least not only. The 
problem is present even for basic operations level like multiplication.

As stated before, a modular multiplication using the Scala 
implementation is at least 30 times slower than the same operation using 
jsbn (using the last version of Chromium). I redid the test with a 
1000-loop instead of 100 and it's even worse: more than a 60 ratio.

In internal the digits are stored into an Int array. Maybe the 
javascript representation of the array produced by Scala.js prevents the 
javascript engine to apply some optimisations. That's just an hypothesis.



Le 04/07/2015 19:11, Alistair Johnson a écrit :
>
> I have a small example test, that can also run in the repl:
>
>        import  java.math.BigInteger
>        import  java.util.Arrays
>
>        import  java.util.Random
>        val  bi1  =  new  BigInteger(1024,new  Random())
>        val  bi2  =  new  BigInteger(1024,new  Random())
>        val  bi3  =  new  BigInteger(1024,new  Random())
>        val  t1  =  scala.compat.Platform.currentTime
>
>        val  b  =  bi1.modPow(bi2,bi3)
>        val  t2  =  scala.compat.Platform.currentTime-  t1
>        println (s"$b ModPow TIME:: ${t2}ms")
>
>        val  t3  =  scala.compat.Platform.currentTime
>        for(i<-  0  until100){
>          val  c  =  bi1.multiply(bi2).mod(bi3)
>        }
>        val  t4  =  scala.compat.Platform.currentTime-  t3
>        println (s"$b Mult mod TIME:: ${t4}ms")
>
> Typical times SJS/JVM are 1000,146 / 133/129
>
> So, on the second test the SJS is marginally slower but the first is 
> /much/ slower. Also, the JVM test here is way slower than the jsbn ones.
>
> Looking at the code, jsbn uses a completely different algorithm for 
> |ModPow|, SJS uses the sliding window method. No doubt the JVM 
> algorithm also differs from the SJS one.
>
> So I guess the solution here is to modify the actual algorithm used
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/scala-js/scala-js/issues/1785#issuecomment-118535101>.
>


_________________________________________________________________________________________________________________________

Ce message et ses pieces jointes peuvent contenir des informations confidentielles ou privilegiees et ne doivent donc
pas etre diffuses, exploites ou copies sans autorisation. Si vous avez recu ce message par erreur, veuillez le signaler
a l'expediteur et le detruire ainsi que les pieces jointes. Les messages electroniques etant susceptibles d'alteration,
Orange decline toute responsabilite si ce message a ete altere, deforme ou falsifie. Merci.

This message and its attachments may contain confidential or privileged information that may be protected by law;
they should not be distributed, used or copied without authorisation.
If you have received this email in error, please notify the sender and delete this message and its attachments.
As emails may be altered, Orange is not liable for messages that have been modified, changed or falsified.
Thank you.

**Comment by InTheNow** [comment]
OK, I'll try to clarify: the code snippet I posted has two tests, and I executed it twice: Once in the repl, using the standard JVM library with its own, internal BigInteger implementation. The main point here is that mult.mod test is about the same as the scalajs one

I.E.  - the _pure_ java tests are slower than the jsbn ones,  completely excluding scala.js and any of its libraries. 

Perhaps you have a link to the exact tests that you have made?

Focusing on the modPow for now, I added a bunch of `println` in the sjs code and it is for sure the sliding window that is culprit. Looking at [wikipedia](https://en.wikipedia.org/wiki/Exponentiation_by_squaring#Sliding_window_method), and also the [jsbn code](https://github.com/andyperlitch/jsbn/blob/master/index.js#L974), it looks to me that jsbn is  using a different method, based on Fixed BAsed exponent and pre-computations. 

So I think it best that we try and compare apples with apples, and hence implement the same, faster algorithms in sjs

**Comment by InTheNow** [comment]
Jackpot: [Here is some scala code](https://github.com/soc/bigint/blob/master/src/main/scala/SBigInt.scala)  that uses a similar approach - @soc Any advice here, please?

**Comment by InTheNow** [comment]
In addition to the link above, it's perhaps appropriate to add that @soc did discuss the code when the scala.js BigNums were being implemented but the goal then was "just" get a full, working implementation of `BigInteger` and `BigDecimal`, and add the cherries later. Looks like that later => now :)

**Comment by ndesmoul** [comment]
No, some precomputation on the base is normal with the sliding window algorithm. It is the way it work.

There are some algorithms specific for fixed-base exponentiation, but whose main interest is for situation where the base is known long in advance. In that case you can do a lot of precomputation and do the actual operation a lot more faster. See FComb in handbook of crypto for example.

Anyway as for the test I did (sorry it will be a little long...):

Here is my wrapper for jsbn (may be incomplete or bugged).

    package js.math.jsbn
    @JSName("BigInteger")
    class BigInt protected () extends js.Object {
      def this(a: Int, c: RandomGenerator) = this()
      def this(a: Int, b: Int, c: RandomGenerator) = this()
      def this(a: String, b: Int) = this()
      def this(a: BigInt) = this()
      def am(i: Double, x: Double, w: BigInt, j: Double, c: Double, n: Double): Double = js.native
      def copyTo(r: BigInt): Unit = js.native
      def fromInt(x: Int): Unit = js.native
      def fromString(x: String, b: Int): Unit = js.native
      def clamp(): Unit = js.native
      def toString(b: Int = ???): String = js.native
      def negate(): BigInt = js.native
      def abs(): BigInt = js.native
      def compareTo(a: BigInt): Int = js.native
      def bitLength(): Int = js.native
      def dlShiftTo(n: Int, r: BigInt): Unit = js.native
      def drShiftTo(n: Int, r: BigInt): Unit = js.native
      def lShiftTo(n: Int, r: BigInt): Unit = js.native
      def rShiftTo(n: Int, r: BigInt): Unit = js.native
      def subTo(a: BigInt, r: BigInt): Unit = js.native
      def multiplyTo(a: BigInt, r: BigInt): Unit = js.native
      def squareTo(r: BigInt): Unit = js.native
      def divRemTo(m: BigInt, q: BigInt, r: BigInt): Unit = js.native
      def mod(a: BigInt): BigInt = js.native
      def invDigit(): Double = js.native
      def isEven(): Boolean = js.native
      def exp(e: Double, z: Reduction): BigInt = js.native
      def modPowInt(e: Double, m: BigInt): BigInt = js.native
      override def clone(): BigInt = js.native
      def intValue(): Int = js.native
      def byteValue(): Byte = js.native
      def shortValue(): Short = js.native
      def chunkSize(r: Double): Double = js.native
      def signum(): Int = js.native
      def toRadix(b: Int): String = js.native
      def fromRadix(s: String, b: Double): Unit = js.native
      def fromNumber(a: Double, b: Double = ???, c: Double = ???): Unit = js.native
      def toByteArray(): js.Array[Byte] = js.native
      //def toByteArray(): Array[Byte] = js.native
      def equals(a: BigInt): Boolean = js.native
      def min(a: BigInt): BigInt = js.native
      def max(a: BigInt): BigInt = js.native
      def bitwiseTo(a: BigInt, op: js.Function2[Double, Double, Double], r: BigInt): Unit = js.native
      def and(a: BigInt): BigInt = js.native
      def or(a: BigInt): BigInt = js.native
      def xor(a: BigInt): BigInt = js.native
      def andNot(a: BigInt): BigInt = js.native
      def not(): BigInt = js.native
      def shiftLeft(n: Int): BigInt = js.native
      def shiftRight(n: Int): BigInt = js.native
      def getLowestSetBit(): Double = js.native
      def bitCount(): Int = js.native
      def testBit(n: Int): Boolean = js.native
      def changeBit(n: Int, op: js.Function2[Double, Double, Double]): BigInt = js.native
      def setBit(n: Int): BigInt = js.native
      def clearBit(n: Int): BigInt = js.native
      def flipBit(n: Int): BigInt = js.native
      def addTo(a: BigInt, r: BigInt): Unit = js.native
      def add(a: BigInt): BigInt = js.native
      def subtract(a: BigInt): BigInt = js.native
      def multiply(a: BigInt): BigInt = js.native
      def square(): BigInt = js.native
      def divide(a: BigInt): BigInt = js.native
      def remainder(a: BigInt): BigInt = js.native
      def divideAndRemainder(a: BigInt): js.Array[BigInt] = js.native
      def dMultiply(n: Double): Unit = js.native
      def dAddOffset(n: Double, w: Double): Unit = js.native
      def pow(e: Int): BigInt = js.native
      def multiplyLowerTo(a: BigInt, n: Double, r: BigInt): Unit = js.native
      def multiplyUpperTo(a: BigInt, n: Double, r: BigInt): Unit = js.native
      def modPow(e: BigInt, m: BigInt): BigInt = js.native
      def gcd(a: BigInt): BigInt = js.native
      def modInt(n: Double): Double = js.native
      def modInverse(m: BigInt): BigInt = js.native
      def isProbablePrime(t: Double): Boolean = js.native
      def millerRabin(t: Double): Boolean = js.native
      def valueOf(x:Long):BigInt = js.native
    }


And now the test:

    def benchmarkBigInteger() = {

    println("Bench BigInteger")
    import java.math.BigInteger
    
    // generate a prime number each time is too long, so here's one:
    val modHexa = "8600638d58155caded48cb45a1258593d775a03fe82679f62c62ec6c75e62e76fcb4e565f74675abe6cf6602939f6b4a159f56843f02f3f825e0f176a8df9f04f41eaa1cf60e598ca892670e30ea5b2d7002fb500b84e331d238708f600a14ec791dec30afc0fad2e5a0338847b0505f6d0a839516c892c06f4015362fbce145"
  
    val rand = new java.util.Random()   // SecureRandom missing (possible to have a good random source with javascript?)
    val aHexa =  new BigInteger(1024, rand).toString(16)
    val bHexa =  new BigInteger(1024, rand).toString(16)
    
    println("Hexa String version")
    println("mod = " + modHexa)
    println("a = " + aHexa)
    println("b = " + bHexa)
    
    // here the BigInt is for the BigInt of jsbn, not the one from the Scala lib
    val modBI = new BigInt(modHexa, 16)
    val aBI = new BigInt(aHexa, 16)
    val bBI = new BigInt(bHexa, 16) 
    
    println("jsbn version")
    println("modBI = " + modBI.toString(16))
    println("aBI = " + aBI.toString(16))
    println("bBI = " + bBI.toString(16))
    
    val mod =  new BigInteger(modHexa, 16)
    val a =  new BigInteger(aHexa, 16)
    val b =  new BigInteger(bHexa, 16)
    
    println("BigInteger version")
    println("mod = " + mod.toString(16))
    println("a = " + a.toString(16))
    println("b = " + b.toString(16))
    
    println("Do modular multiplication")
       
    var t0, t1:Long = 0
    val nbLoop = 1000
    var mutRes:BigInteger = null
    t0 = System.currentTimeMillis()
    for(i <- 0 until nbLoop) {
      mutRes = a multiply b mod mod
    }
    t1 = System.currentTimeMillis()
    val tBigInteger = (t1 - t0) * 1.0 / nbLoop 
    println("BigInteger result = " + mutRes.toString(16))
    println("" + tBigInteger + " ms")
    
    var mutBIRes:BigInt = null
    t0 = System.currentTimeMillis()
    for(i <- 0 until nbLoop) {
      mutBIRes = aBI multiply bBI mod modBI
    }
    t1 = System.currentTimeMillis()
    val tJsbn = (t1 - t0) * 1.0 / nbLoop
    
    println("JSBN result      = " + mutBIRes.toString(16))
    println("" + tJsbn + " ms")
      
    // modpow test code is almost the same, so not shown here...
    }
    

**Timing results of modular multiplication on Chromium**:
* modular mutiplication
   * BigInteger (Scala version): 1.1671 ms 
   * JSBN: 0.0212 ms
* modpow:
   * BigInteger (Scala version):  735 ms
   * JSBN: 18 ms

**Some timing results under OpenJDK 8**:
With BigInteger (from standard lib) :
* modular multiplication: 0.0042 ms
* modPow: 2.42 ms

So JVM is faster (well no surprise here...). And when executed as javascript, JSBN is a lot faster even for modular multiplication (by a factor 55 here).

Still on JVM I tested the *Scala BigInteger of scala.js*:
* modular multiplication: 0.0068 ms
* modPow: 5.79 ms
Not very good compared to the standard lib as it is almost 2 times slower.

But I have the "original" java code used by scala.js, and surprisingly the performances are not the same: 
* modular multiplication: 0.0039 ms 
* modPow: 2.51 ms
So almost the same than the BigInteger from the standard lib.

This post is already too big, but here are some conclusion :
- The Scala.js BigInteger is almost 2 times slower than the original java code it is converted from, which I find a little strange. 
- The same Scala code executed as javascript is horribly slower than jsbn (more than 40 slower).
- ModPow is not the only operation affected, as the modular multiplication is also very (very!) slow once in javascript.

**Comment by sjrd** [comment]
> The Scala.js BigInteger is almost 2 times slower than the original java code it is converted from, which I find a little strange. 

Not strange. Some of it was refactored to more idiomatic Scala code when I could vouch for the Scala.js optimizer to be able to optimize the abstractions away. The Scala/JVM optimizer is not that good, so it would produce sub-optimal code on the JVM compared to the Java code it came from.

> The same Scala code executed as javascript is horribly slower than jsbn (more than 40 slower).

Which clearly means we should take lessons/code (if possible according to the license) from jsbn to make the Scala.js version as fast (or faster) than jsbn.

**Comment by soc** [comment]
The BigInt implementation I was hacking on some time ago is pretty much complete (I think the version on GitHub is quite old), except for a few missing methods.

Please note that the algorithms aren't some code I came up with, but code licensed under a permissive license from some of the best experts on this topic, Tim Buktu and Alan Eliasen, and are heavily reviewed by more experts: https://github.com/tbuktu/bigint/commit/df5f3ed717420d21fc8e68cf5344bbef470dbedd

It's basically the same code that's replacing the old `java.math.BigInteger` implementation in Java 8/Java 9/...

The Apache Harmony approach feels like they cared more about explicitly doing things as different as possible so that people don't start to wonder whether they copied stuff from the JDK, and cared less about writing good, efficient code.

I'll search for the most recent version of this implementation and push it to GitHub, then everyone can decide whether he/she wants to push the implementation across the finish line.

**Comment by ndesmoul** [comment]
@sjrd, I understand your argument for the bad performances of the Scala BigInteger conversion under a JVM. (I guess trying to have a more Scala like code at this low level is not a good idea).

But does it really explain the integrality of the big gap with jsbn? I mean, if that was the case, the ratio should be 2 or 3 instead of 10 times more. After all the Apache Harmony BigInteger doesn't perform so bad compared to the BigInteger of Java8, which itself is probably on par with jsbn algorithmically.
Am I missing something?

As @soc said though, the Harmony code is quite complicated, and it may be possible to do better with fewer code. Maybe his implementation will provide the solution.

**Comment by soc** [comment]
@ndesmoul I wouldn't say that the code can be considers complicated when compared to https://github.com/tbuktu/bigint, it's just very very slow.

Have a look at the benchmarks. Their BigInteger implementation *destroys* the one that shipped with OpenJDK7. It depends on how bad it gets, with _number of digits_ = 1000000, it's 23 seconds vs. 0.16 seconds, more than two orders of magnitude.

I expect that the Harmony code runs considerably worse than the OpenJDK7 one.

More performance requires more code, because different, increasingly complex algorithms become worthwhile when the number of digits increases.

**Comment by soc** [comment]
Ok, I think the most recent implementation should be here: https://github.com/soc/bigint

I'm still looking whether I have some of the missing bits and pieces implemented somewhere already...

**Comment by sjrd** [comment]
The improvements on `Long`s from #2488 have brought significant improvements to `BigInteger`s, with some 8x speedups on some operations, notably `modPow`. Further (smaller) improvements are coming in #2544 and should land in master before 0.6.11 is released.

**Comment by sjrd** [comment]
This was basically fixed in 0.6.11. It's not the most performance implementation ever, but at least it's not *terrible* anymore. Further improvements are welcome, obviously, but I'll close this issue.
